rust/otap-dataflow/crates/otlp/src/otlp_batch_processor.rs

```diff
@@ -10,134 +10,154 @@ use otap_df_engine::error::Error;
-                            current_batch.resource_spans.push(res.clone());
-                            // Clear res.scope_spans after pushing to prevent duplicating scope groups in subsequent batches.
-                            // This ensures each batch only contains the intended scope groups for that batch.
-                            res.scope_spans.clear();
-                        }
-                        // Push the full batch to output and start a new one.
-                        batches.push(current_batch.clone());
+                        current_batch.resource_spans.push(res);
+                        batches.push(current_batch);
+
+                        // start a fresh batch / resource container
                         current_batch = ExportTraceServiceRequest {
                             resource_spans: Vec::new(),
                         };
                         current_span_count = 0;
+                        res = ResourceSpans {
+                            resource: rs.resource.clone(),
+                            scope_spans: Vec::new(),
+                            schema_url: rs.schema_url.clone(),
+                        };
                     }
                 }
             }
+
+            // still have some scope-spans for this resource → keep them for the
+            // next iteration or for the final flush
             if !res.scope_spans.is_empty() {
-                current_batch.resource_spans.push(res.clone());
+                current_batch.resource_spans.push(res);
             }
         }
-        if current_span_count > 0 && !current_batch.resource_spans.is_empty() {
+
+        // flush the last (possibly not-full) batch
+        if !current_batch.resource_spans.is_empty() {
             batches.push(current_batch);
         }
-        batches
+
+        Ok(batches)
     }
 }
 
 impl HierarchicalBatchSplit for ExportMetricsServiceRequest {
-    fn split_into_batches(mut self, max_batch_size: usize) -> Vec<Self> {
-        let mut batches = Vec::new();
-
-        // The batch currently being filled; pushed to output when max_batch_size is reached.
-        let mut current_batch = ExportMetricsServiceRequest::default();
-        let mut current_metric_count = 0;
+    fn split_into_batches(mut self, max_batch_size: usize) -> Result<Vec<Self>, Error<OTLPData>> {
+        if max_batch_size == 0 {
+            return Err(Error::ProcessorError {
+                processor: Cow::Borrowed("HierarchicalBatchSplit::ExportMetricsServiceRequest"),
+                error: "max_batch_size must be greater than zero".into(),
+            });
+        }
+        let total_metrics = self
+            .resource_metrics
+            .iter()
+            .flat_map(|rm| &rm.scope_metrics)
+            .map(|sm| sm.metrics.len())
+            .sum::<usize>();
+        let estimated_batches = total_metrics.div_ceil(max_batch_size);
+        let mut batches = Vec::with_capacity(estimated_batches);
```

Comment:
It worries me to see three copies of this logic--they may have diverged a bit. I would like to see the next step be to create a generic version of this, I also suggest merging what we have.